src/libcommonserver/io/iohelper_win.cpp

```diff
@@ -329,126 +331,296 @@ bool IoHelper::checkIfFileIsDehydrated(const SyncPath &itemPath, bool &isDehydra
     return IoHelper::getXAttrValue(itemPath.native(), FILE_ATTRIBUTE_OFFLINE, isDehydrated, ioError);
 }
 
+static bool setRightsWindowsApi(const SyncPath &path, DWORD permission, ACCESS_MODE accessMode, IoError &ioError,
+                                log4cplus::Logger logger) noexcept {  // Return false if we should try the fallback method
+
+    PACL pACL_old = nullptr;  // Current ACL
+    PACL pACL_new = nullptr;  // New ACL
+    PSECURITY_DESCRIPTOR pSecurityDescriptor = nullptr;
+    EXPLICIT_ACCESS ExplicitAccess;
+    ZeroMemory(&ExplicitAccess, sizeof(ExplicitAccess));
+
+    ExplicitAccess.grfAccessPermissions = permission;
+    ExplicitAccess.grfAccessMode = accessMode;
+    ExplicitAccess.grfInheritance = NO_INHERITANCE;
+    ExplicitAccess.Trustee.pMultipleTrustee = nullptr;
+    ExplicitAccess.Trustee.MultipleTrusteeOperation = NO_MULTIPLE_TRUSTEE;
+    ExplicitAccess.Trustee.TrusteeForm = Utility::_trustee.TrusteeForm;
+    ExplicitAccess.Trustee.TrusteeType = Utility::_trustee.TrusteeType;
+    ExplicitAccess.Trustee.ptstrName = Utility::_trustee.ptstrName;
+
+    std::wstring path_wstr = Path2WStr(path);
+    size_t pathw_len = path_wstr.length();
+
+    auto pathw_ptr = std::make_unique<WCHAR[]>(pathw_len + 1);
+    path_wstr.copy(pathw_ptr.get(), pathw_len);
+
+    LPCWSTR pathw_c = path_wstr.c_str();
+    LPWSTR pathw = pathw_ptr.get();
+    pathw[pathw_len] = L'\0';
+
+    DWORD ValueReturned = GetNamedSecurityInfo(pathw_c, SE_FILE_OBJECT, DACL_SECURITY_INFORMATION, nullptr, nullptr, &pACL_old,
+                                               nullptr, &pSecurityDescriptor);
+    ioError = dWordError2ioError(ValueReturned);
+    if (ValueReturned != ERROR_SUCCESS) {
+        ioError = dWordError2ioError(ValueReturned);
+        LOGW_WARN(logger, L"Error in GetNamedSecurityInfo: " << Utility::formatIoError(path, ioError).c_str()
+                                                             << L" | DWORD error: " << ValueReturned);
+        LocalFree(pSecurityDescriptor);
+        LocalFree(pACL_new);
+        // pACL_old is a pointer to the ACL in the security descriptor, so it should not be freed.
+        if (ioError == IoErrorNoSuchFileOrDirectory || ioError == IoErrorAccessDenied) {
```

Comment:
Use _isExpectedError

Reply:
setRightsWindowsApi is not a member function of IoHelper (and so can not access `_isExpectedError`).

Reply:
https://github.com/Infomaniak/desktop-kDrive/pull/68#discussion_r1583012018